test/playlist-loader.test.js

```diff
@@ -469,1593 +423,2109 @@ QUnit.test('updateMaster changes old values', function(assert) {
+      '#EXT-X-SERVER-CONTROL:CAN-BLOCK-RELOAD=YES\n' +
+      '#EXTINF:2\n' +
+      'segment0.ts\n' +
+      '#EXTINF:2\n' +
+      'segment1.ts\n' +
+      '#EXTINF:2\n' +
+      'segment2.ts\n' +
+      '#EXTINF:2\n' +
+      'segment3.ts\n' +
+      '#EXTINF:2\n' +
+      'segment4.ts\n' +
+      '#EXTINF:2\n' +
+      'segment5.ts\n' +
+      '#EXT-X-PART:URI="segment6-part1.ts",DURATION=1\n' +
+      '#EXT-X-PART:URI="segment6-part2.ts",DURATION=1\n' +
+      'segment6.ts\n' +
+      '#EXT-X-PART:URI="segment7-part1.ts",DURATION=1\n' +
+      '#EXT-X-PART:URI="segment7-part2.ts",DURATION=1\n' +
+      'segment7.ts\n' +
+      '#EXT-X-PART:URI="segment8-part1.ts",DURATION=1\n' +
+      '#EXT-X-PART:URI="segment8-part2.ts",DURATION=1\n' +
+      'segment8.ts\n'
     );
-    assert.ok(loader.media().endList, 'flushed the final line of input');
-  }
-);
 
-QUnit.test('Supports multiple STREAM-INF with the same URI', function(assert) {
-  const loader = new PlaylistLoader('master.m3u8', this.fakeVhs);
+    loader.trigger('mediaupdatetimeout');
 
-  loader.load();
+    assert.equal(this.requests[0].uri, 'http://example.com/media.m3u8?_HLS_msn=9');
+  });
 
-  this.requests.shift().respond(
-    200, null,
-    '#EXTM3U\n' +
-                                '#EXT-X-STREAM-INF:BANDWIDTH=1,AUDIO="aud0"\n' +
-                                'video/media.m3u8\n' +
-                                '#EXT-X-STREAM-INF:BANDWIDTH=2,AUDIO="aud1"\n' +
-                                'video/media.m3u8\n'
-  );
-  assert.equal(
-    loader.master.playlists['0-video/media.m3u8'].id,
-    loader.master.playlists[0].id,
-    'created key based on playlist id'
-  );
+  QUnit.test('can add all query directives', function(assert) {
+    this.fakeVhs.options_ = {experimentalLLHLS: true};
+    const loader = new PlaylistLoader('http://example.com/media.m3u8', this.fakeVhs);
 
-  assert.equal(
-    loader.master.playlists['1-video/media.m3u8'].id,
-    loader.master.playlists[1].id,
-    'created key based on playlist id'
-  );
+    loader.load();
+
+    // no newline
+    this.requests.shift().respond(
+      200, null,
+      '#EXTM3U\n' +
+      '#EXT-X-MEDIA-SEQUENCE:0\n' +
+      '#EXT-X-SERVER-CONTROL:CAN-BLOCK-RELOAD=YES,CAN-SKIP-UNTIL=3\n' +
+      '#EXTINF:2\n' +
+      'segment0.ts\n' +
+      '#EXTINF:2\n' +
+      'segment1.ts\n' +
+      '#EXTINF:2\n' +
+      'segment2.ts\n' +
+      '#EXTINF:2\n' +
+      'segment3.ts\n' +
+      '#EXTINF:2\n' +
+      'segment4.ts\n' +
+      '#EXTINF:2\n' +
+      'segment5.ts\n' +
+      '#EXT-X-PART:URI="segment6-part1.ts",DURATION=1\n' +
+      '#EXT-X-PART:URI="segment6-part2.ts",DURATION=1\n' +
+      'segment6.ts\n' +
+      '#EXT-X-PART:URI="segment7-part1.ts",DURATION=1\n' +
+      '#EXT-X-PART:URI="segment7-part2.ts",DURATION=1\n' +
+      'segment7.ts\n' +
+      '#EXT-X-PART:URI="segment8-part1.ts",DURATION=1\n' +
+      '#EXT-X-PRELOAD-HINT:TYPE="PART",URI="segment8-part2.ts"\n'
```

Comment:
Non blocking, but since these tests use a pretty standard form, it might be worth creating a function to generate the manifest strings (num segments, num parts, preload hints, etc.).

Reply:
It might be nice to do that, but I think that ultimately it will just be a json representation of what you want that has to be built into a string, instead of just building the string directly.